Source/Core/Common/Arm64Emitter.h

```diff
@@ -558,156 +534,44 @@ struct LogicalImm
-        0x0101010101010101UL,
-        0x1111111111111111UL,
-        0x5555555555555555UL,
-    }};
-
-    const int multiplier_idx = Common::CountLeadingZeros((u64)d) - 57;
-
-    // Ensure that the index to the multipliers array is within bounds.
-    DEBUG_ASSERT((multiplier_idx >= 0) &&
-                 (static_cast<size_t>(multiplier_idx) < multipliers.size()));
-
-    const u64 multiplier = multipliers[multiplier_idx];
-    const u64 candidate = (b - a) * multiplier;
-
-    // The candidate pattern doesn't match our input value, so fail.
-    if (value != candidate)
-      return;
+    const size_t rotation = Common::CountTrailingZeros(value & (value + 1));
+    const u64 normalized = Common::RotateRight(value, rotation);
 
-    // We have a match! This is a valid logical immediate, so now we have to
-    // construct the bits and pieces of the instruction encoding that generates
-    // it.
-    n = out_n;
+    const size_t esize = Common::CountTrailingZeros(normalized & (normalized + 1));
+    const size_t ones = Common::CountTrailingZeros(~normalized);
 
-    // Count the set bits in our basic stretch. The special case of clz(0) == -1
-    // makes the answer come out right for stretches that reach the very top of
-    // the word (e.g. numbers like 0xffffc00000000000).
-    const int clz_b = (b == 0) ? -1 : Common::CountLeadingZeros(b);
-    s = clz_a - clz_b;
+    // Check the value is repeating; also ensures esize is a power of two.
 
-    // Decide how many bits to rotate right by, to put the low bit of that basic
-    // stretch in position a.
-    if (negate)
-    {
-      // If we inverted the input right at the start of this function, here's
-      // where we compensate: the number of set bits becomes the number of clear
-      // bits, and the rotation count is based on position b rather than position
-      // a (since b is the location of the 'lowest' 1 bit after inversion).
-      s = d - s;
-      r = (clz_b + 1) & (d - 1);
-    }
-    else
+    if (Common::RotateRight(value, esize) != value)
     {
-      r = (clz_a + 1) & (d - 1);
+      valid = false;
+      return;
     }
 
-    // Now we're done, except for having to encode the S output in such a way that
+    // Now we're done. We just have to encode the S output in such a way that
     // it gives both the number of set bits and the length of the repeated
-    // segment. The s field is encoded like this:
-    //
-    //     imms    size        S
-    //    ssssss    64    UInt(ssssss)
-    //    0sssss    32    UInt(sssss)
-    //    10ssss    16    UInt(ssss)
-    //    110sss     8    UInt(sss)
-    //    1110ss     4    UInt(ss)
-    //    11110s     2    UInt(s)
-    //
-    // So we 'or' (-d << 1) with our computed s to form imms.
-    s = ((-d << 1) | (s - 1)) & 0x3f;
+    // segment.
+
+    const size_t tmp = ((~esize + 1) << 1) | (ones - 1);
+
+    r = static_cast<u8>((esize - rotation) & (esize - 1));
+    s = tmp & 0x3f;
+    n = (~tmp >> 6) & 1;
```

Comment:
Probably more performant:

```suggestion
    n = (esize >> 6) & 1;
```